cocotb/triggers.py

```diff
@@ -917,7 +937,22 @@ async def with_timeout(trigger, timeout_time, timeout_unit="step"):
 
     .. deprecated:: 1.5
         Using None as the the *timeout_unit* argument is deprecated, use ``'step'`` instead.
-   """
+
+    .. versionchanged:: 1.6
+        Support passing :term:`python:coroutine`\ s.
+
+    .. versionchanged:: 2.0
+        :class:`~cocotb.coroutine`\ s that are passed unstarted are now killed if there is a timeout.
+    """
+
+    if inspect.iscoroutine(trigger) or isinstance(trigger, cocotb.decorators.coroutine) and not trigger._started:
```

Comment:
I think we need `RunningCoroutine` here, otherwise it just skips to the normal behavior.
```suggestion
    if inspect.iscoroutine(trigger) or isinstance(trigger, cocotb.decorators.RunningCoroutine) and not trigger._started:
```